kevm-pyk/src/kevm_pyk/kproj/evm-semantics/abi.md

```diff
@@ -510,21 +513,34 @@ where `F1 : F2 : F3 : F4` is the (two's complement) byte-array representation of
 
     rule #isStaticType(#array(_, _, _)) => false
 
+    rule #isStaticType(#tuple( ARGS )) => notBool #hasDynamicType(ARGS)
+
+    syntax Bool ::= #hasDynamicType(TypedArgs) [function, total]
+ // ------------------------------------------------------------
+    rule #hasDynamicType(.TypedArgs) => false
+    rule #hasDynamicType(T, TS) => #hasDynamicType(TS) requires #isStaticType(T)
+    rule #hasDynamicType(T,  _) => true requires notBool #isStaticType(T)
+
     syntax Int ::= #sizeOfDynamicType ( TypedArg ) [function]
  // ---------------------------------------------------------
     rule #sizeOfDynamicType(#bytes(BS)) => 32 +Int #ceil32(lengthBytes(BS))
 
+    rule #sizeOfDynamicType(#tuple(ARGS)) => 32 +Int #sizeOfDynamicTypeAux(ARGS)
+
     rule #sizeOfDynamicType(#array(T, N, _)) => 32 *Int (1 +Int N)
       requires #isStaticType(T)
 
     rule #sizeOfDynamicType(#array(T, N, ELEMS)) => 32 *Int (1 +Int N +Int #sizeOfDynamicTypeAux(ELEMS))
       requires notBool #isStaticType(T)
 
-    syntax Int ::= #sizeOfDynamicTypeAux ( TypedArgs ) [function]
- // -------------------------------------------------------------
+    syntax Int ::= #sizeOfDynamicTypeAux ( TypedArgs ) [function, total]
+ // --------------------------------------------------------------------
     rule #sizeOfDynamicTypeAux(TARG, TARGS) => #sizeOfDynamicType(TARG) +Int #sizeOfDynamicTypeAux(TARGS)
```

Comment:
`#sizeOfDynamicTypeAux` is total if and only if `#sizeOfDynamicType` is total.

Reply:
Right, so it's not total in this case. I'll fix this.

Reply:
Actually, I think it's a bit more interesting than that. The rule fires only when `notBool #isStaticType(TARG)`, which has the `#string` case that's not covered in `#sizeOfDynamicType`. Could you create a case for `#string` there? If so, then `sizeOfDynamicTypeList` will be total.

Reply:
Didn't notice that `#sizeOfDynamicType` is missing a rule for `#string`. I'll add that!